### DISCUSSION TITLE
Automatically Solving the N+1 Problem with `FindUnique`

### DISCUSSION THREAD

**Discussion by KoichiKiyokawa** [prisma/prisma-client-go#904] [Q&A] [answered]
Firstly, I'd like to express my appreciation for all the work you've put into this library 🎉 

The Prisma TypeScript client automatically addresses the N+1 problem when we use the `findUnique` method, as detailed here: https://www.prisma.io/docs/guides/performance-and-optimization/query-optimization-performance#solving-n1-in-graphql-with-findunique-and-prismas-dataloader

Does the prisma-client-go offer a similar solution to the N+1 problem like the TypeScript client does? I attempted to implement the following code, but it resulted in an N+1 problem:
```go
func (r *userResolver) Posts(ctx context.Context, obj *model.User) ([]*model.Post, error) {
	u, err := r.db.User.FindUnique(db.User.ID.Equals(obj.ID)).With(db.User.Posts.Fetch()).Exec(ctx)
```
(Repro is here https://github.com/KoichiKiyokawa/prisma-client-go-nplusone-repro)

Thanks.

**Comment by steebchen** [comment, marked as answer]
Thanks!

The Go client currently does not do any batching at this time. I will have to think about this. I'm wondering if this is actually a good solution as it doesn't really solve the n+1 problem universally, it just solves it for one use case (when you do multiple findUnique) – please correct me if that's wrong though.

For now, I would recommend doing `.Fetch()` for whatever data you most like need in the parent resolver.

For example, if you are able to fetch `user -> user's posts -> post's comments`, you would run into the n+1 problem. 

For example:

```go
func (r *Viewer) User(ctx context.Context, obj *models.Viewer, id string) (*db.UserModel, error) {
	return r.db.User.FindFirst(
		db.User.ID.Equals(id),
	).With(
		db.User.Posts.Fetch().With(
			db.Post.Author.Fetch(),
		),
	).Exec(ctx)
}
```

Of course this would only solve it for that query and not universally, but I'm not sure what the best solution is here
